worlds/witness/rules.py

```diff
@@ -0,0 +1,179 @@
+                            logic.EVENT_ITEM_NAMES[item], player
+                        )
+                    else:
+                        valid_option = self.can_reach(
+                            logic.CHECKS_BY_HEX[item]["checkName"],
+                            "Location", player)
+                    if not valid_option:
+                        break
+                elif not self.has(item, player):
+                    valid_option = False
+                    break
+
+            if valid_option:
+                return True
+
+        return False
+
+    def _safe_manual_panel_check(self, panel, world, player, logic, locat):
+        """
+        nested can_reach can cause problems, but only if the region being
+        checked is neither of the two original regions from the first
+        can_reach.
+        A nested can_reach is okay here because the only panels this
+        function is called on are panels that exist on either side of all
+        connections they are required for.
+        The spoiler log looks so much nicer this way,
+        it gets rid of a bunch of event items, only leaving a couple. :)
+        """
+        region = logic.CHECKS_BY_HEX[panel]["region"]["name"]
+
+        return (
+            self.meets_item_requirements(panel, world, player, logic, locat)
+            and self.can_reach(region, "Region", player)
+        )
+
+    def can_solve_panels(self, panel_hex_to_solve_set, world, player, logic, locat):
+        """
+        Checks whether a set of panels can be solved.
+        """
+
+        for option in panel_hex_to_solve_set:
+            if len(option) == 0:
+                return True
+
+            valid_option = True
+
+            for panel in option:
+                if not self._can_solve_panel(panel, world, player, logic, locat):
+                    valid_option = False
+                    break
+
+            if valid_option:
+                return True
+        return False
+
+
+def make_lambda(check_hex, world, player, logic, locat):
+    """
+    Lambdas are created in a for loop so values need to be captured
+    """
+    return lambda state: state.meets_item_requirements(
+        check_hex, world, player, logic, locat
+    )
+
+
+def set_rules(world: MultiWorld, player: int,
+              logic: WitnessLogic, locat: WitnessLocations):
+    """
+    Sets all rules for all locations
+    """
+
+    for location in locat.CHECK_LOCATION_TABLE:
+        real_location = location
+
+        if location in locat.EVENT_LOCATION_TABLE:
+            real_location = location[:-7]
+
+        panel = logic.CHECKS_BY_NAME[real_location]
+        check_hex = panel["checkHex"]
+
+        rule = make_lambda(check_hex, world, player, logic, locat)
+
+        set_rule(world.get_location(location, player), rule)
+
+    if world.logic[player] != 'nologic':
+        world.completion_condition[player] = \
+            lambda state: state.has('Victory', player)
```

Comment:
can you have a slot without completion_condition? should probably be added to to api.md then

Reply:
I don't remember where this came from, but it's a case of "It was there before". Idk from whom, or why. I can remove it's not necessary

Reply:
@Berserker66 think you are most qualified to answer my question